lib/nixd/include/nixd/AST.h

```diff
@@ -8,6 +8,7 @@
 #include <llvm/ADT/FunctionExtras.h>
 #include <llvm/ADT/StringRef.h>
 
+#include <nix/config.h>
```

Comment:
Our headers is **sorted** in lexicographic order which basically assumes all headers are self-contained. However, `<nix/config.h>` must be included before any "real" nix headers because it contains many critical macro definition, and place it after a nix header file may lead to incorrect struct layout. (If I'm wrong please comment)

```suggestion
#include <nix/config.h>

```

Let's leave a blank line here, to make sure it is not sorted with the rest nix headers?

---

Also, may I ask why should we do this in C++ source files? (As comment above, `<nix/config.h>` is semantically differently from the rest.

Reply:
I wanted to do it in the cpp source files as that would be resolved normally by the compiler and we don't have to mess around with compiler include ptahs.

In any case I think that this should probably be considered a bug in the Nix headers.

We could just fix the issue there and then drop the `-include` at a later point.

Reply:
> In any case I think that this should probably be considered a bug in the Nix headers.

Frankly Eelco did the same hack here:

https://github.com/edolstra/dwarffs/blob/1f850df9c932acb95da2f31b576a8f6c7c188376/flake.nix#L24

So this might be a "feature" for nix itself?